File: csharp/BCRand/BCRand/Xoshiro256StarStar.cs
using System.Buffers.Binary;
using System.Numerics;

namespace BlockchainCommons.BCRand;

/// <summary>
/// Internal implementation of the Xoshiro256** PRNG algorithm.
/// Used by <see cref="SeededRandomNumberGenerator"/> for deterministic output.
/// </summary>
internal sealed class Xoshiro256StarStar
{
    private ulong _s0;
    private ulong _s1;
    private ulong _s2;
    private ulong _s3;

    /// <summary>
    /// Initializes the PRNG from a 32-byte seed (4 little-endian 64-bit unsigned integers).
    /// </summary>
    public Xoshiro256StarStar(ReadOnlySpan<byte> seed)
    {
        if (seed.Length != 32)
            throw new ArgumentException("Seed must be exactly 32 bytes.", nameof(seed));

        _s0 = BinaryPrimitives.ReadUInt64LittleEndian(seed);
        _s1 = BinaryPrimitives.ReadUInt64LittleEndian(seed[8..]);
        _s2 = BinaryPrimitives.ReadUInt64LittleEndian(seed[16..]);
        _s3 = BinaryPrimitives.ReadUInt64LittleEndian(seed[24..]);
    }

    public ulong NextUInt64()
    {
        ulong result = BitOperations.RotateLeft(_s1 * 5, 7) * 9;
        ulong t = _s1 << 17;

        _s2 ^= _s0;
        _s3 ^= _s1;
        _s1 ^= _s2;
        _s0 ^= _s3;

        _s2 ^= t;
        _s3 = BitOperations.RotateLeft(_s3, 45);

        return result;
    }
}

File: typescript/bc-rand/src/xoshiro256starstar.ts
const MASK64 = 0xFFFFFFFFFFFFFFFFn;

function rotl(x: bigint, k: number): bigint {
    return ((x << BigInt(k)) | (x >> BigInt(64 - k))) & MASK64;
}

/**
 * Internal implementation of the Xoshiro256** PRNG algorithm.
 * Not part of the public API.
 */
export class Xoshiro256StarStar {
    #s: [bigint, bigint, bigint, bigint];

    constructor(s0: bigint, s1: bigint, s2: bigint, s3: bigint) {
        this.#s = [s0 & MASK64, s1 & MASK64, s2 & MASK64, s3 & MASK64];
    }

    nextU64(): bigint {
        const s = this.#s;
        const result = (rotl((s[1] * 5n) & MASK64, 7) * 9n) & MASK64;
        const t = (s[1] << 17n) & MASK64;

        s[2] ^= s[0];
        s[3] ^= s[1];
        s[1] ^= s[2];
        s[0] ^= s[3];
        s[2] ^= t;
        s[3] = rotl(s[3], 45);

        return result;
    }
}
